src/System.Buffers.Experimental/System/Buffers/ReadOnlyBytes.cs

```diff
@@ -329,5 +349,17 @@ public static ReadOnlyBytes Create(params byte[][] buffers)
                 return new ReadOnlyBytes(first);
             }
         }
+
+        public struct Cursor
```

Comment:
Perhaps this should this be
```csharp
public struct Cursor<TBufferList> where TBufferList : IReadOnlyBufferList<byte>
{
    TBufferList _node;
```

Reply:
Then the `CursorOf` method will be generic too, and `this` should be convertible to `TBufferList` which is impossible.

```
        public Cursor<TBufferList> CursorOf<TBufferList>(byte value)
            where TBufferList : IReadOnlyBufferList<byte>
        {
            ReadOnlySpan<byte> first = _first.Span;
            int index = first.IndexOf(value);
            if (index != -1) return new Cursor<TBufferList>(this, index); // CS1503: cannot convert from 'ReadOnlyBytes' to 'T'
            if (_rest == null) return default;
            return CursorOf<TBufferList>((T)_rest, value); // runtime error if _rest is not T
        }
```

Reply:
Its a concrete type at this point not a generic; so would be:
```csharp
public Cursor<ReadOnlySpan<byte>> CursorOf(byte value)
{
    ReadOnlySpan<byte> first = _first.Span;
    int index = first.IndexOf(value);
    if (index != -1) return new Cursor(this, index);
    if (_rest == null) return default;
    return CursorOf(_rest, value); 
}

private static Cursor<ReadOnlySpan<byte>> CursorOf(IReadOnlyBufferList<byte> list, byte value)
{
    ReadOnlySpan<byte> first = list.First.Span;
    int index = first.IndexOf(value);
    if (index != -1) return new Cursor(new ReadOnlyBytes(first, list.Rest), index);
    if (list.Rest == null) return default;
    return CursorOf(list.Rest, value);
}
```
Probably could event drop the generic type from Cursor at that point?

Reply:
Do you mean `ReadOnlyBytes` instead of `ReadOnlySpan<byte>`?

Reply:
Yes :)

Reply:
Coding on a phone is hard :(

Reply:
So if I understand you, you want to move `Cursor` outside of `ReadOnlyBytes` and make it generic. Then a consumer method will accept `Cursor<ReadOnlyBytes>` as an argument like that:

```csharp
public ReadOnlyBytes Slice(Cursor<ReadOnlyBytes> from, Cursor<ReadOnlyBytes> to)
```

Yes? I think it could be simplified.

```csharp
public struct ReadOnlyBytes : IReadOnlyBufferList<byte>
{
    public Cursor CursorOf(byte value)
    {
        ReadOnlySpan<byte> first = _first.Span;
        int index = first.IndexOf(value);
        if (index != -1) return new Cursor(_first, _rest, index);
        if (_rest == null) return default;
        return CursorOf(_rest, value);
    }

    private static Cursor CursorOf(IReadOnlyBufferList<byte> list, byte value)
    {
        ReadOnlySpan<byte> first = list.First.Span;
        int index = first.IndexOf(value);
        if (index != -1) return new Cursor(list.First, list.Rest, index);
        if (list.Rest == null) return default;
        return CursorOf(list.Rest, value);
    }

    public ReadOnlyBytes Slice(Cursor from, Cursor to)
    {
        // if from.First equals to to.First
        //     return from.First.Slice(from.Index, to.Index)
        // first ROB = from.First.Slice(from.Index)
        // while current ROBList not equals to to.Rest
        //     current ROBs = current ROBList.First
        // if to.First is not empty
        //     last ROB = if to.First is not empty
    }
}

public readonly struct Cursor
{
    internal readonly ReadOnlyMemory<byte> First;
    internal readonly IReadOnlyBufferList<byte> Rest;
    internal readonly int Index;

    public Cursor(ReadOnlyMemory<byte> node, IReadOnlyBufferList<byte> rest, int index) { }
}
```